src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -556,7 +556,7 @@ FinalizeOperationNeedingCleanupOnSuccess(const char *operationName)
 	CleanupRecord *record = NULL;
 	foreach_ptr(record, currentOperationRecordList)
 	{
-		if (record->policy == CLEANUP_ALWAYS)
+		if (record->policy == CLEANUP_ALWAYS || record->policy == CLEANUP_DEFERRED_ON_SUCCESS)
```

Comment:
I think that we should only remove resources of type `CLEANUP_ALWAYS` here as we cannot be sure about the success of the operation yet. (e.g. `ROLLBACK` at the end of a transaction makes our prejudgement wrong)

We clean resources of type `CLEANUP_DEFERRED_ON_SUCCESS` at maintenance daemon at configured intervals. Hence, we should not clean resources of type `CLEANUP_ALWAYS` here.

Reply:
We always defer the resource cleanup for `CLEANUP_DEFERRED_ON_SUCCESS` objects since the PR #6258 but current function name containing `...CleanupOnSuccess` confuses me on which objects should we clean there. 

Maybe we can change the method name with `FinalizeOperationNeedingCleanup(const char *operationName, List * policies)` to be more precise on which types of resources we should clean.

Reply:
I think the `...CleanupOnSuccess` means that it's the function we call in case of success, not the objects we clean there. The other function (CleanupOnFailure) is the one we call in case of failure.

Reply:
In both functions, we cleanup objects that are markes as `CLEANUP_ALWAYS`. Additionally, we cleanup objects marked as `CLEANUP_ON_FAILURE` in the failure function as well